Release2.0

lnagle opened the pull request.

lnagle committed:
Refactor method names and remove extraneous function

lnagle committed:
Add callback to stopListening

lnagle committed:
Add stopListening callback test

lnagle committed:
Update ddq version

lnagle committed:
Code Review Changes

lnagle committed:
Add callback to startListening

lnagle committed:
Fix package name

lnagle committed:
In progress changes

lnagle committed:
Update documentation

lnagle committed:
Add ddq 2.0 as peer dependency

lnagle committed:
Minor fixes

AbsentSemicolon commented on script/test-race-conditions.spec.js:
```diff
@@ -0,0 +1,378 @@
+    if (!methods || !methods.length) {
+        doneCb();
+
+        return;
```
Do we need a return here if we call the callback? Don't remember how that works.

lnagle replied:
It is needed to prevent the code below the if block from running.

AbsentSemicolon replied:
Excellent, just wondering.

AbsentSemicolon commented on script/test-race-conditions.spec.js:
```diff
@@ -0,0 +1,378 @@
+ *
+ * Use this when you expect that the method will fail.
+ *
+ * Example: methodFails("nameOfMethod", "param1", "param2", "param3");
```
Should use JsDoc to denote example.

http://usejsdoc.org/tags-example.html

lnagle replied:
Didn't know about that, thanks. Fixed.

AbsentSemicolon commented on script/test-race-conditions.spec.js:
```diff
@@ -0,0 +1,378 @@
+ *
+ * Use this when you expect that the method will succeed.
+ *
+ * Example: methodSucceeds("nameOfMethod", "param1", "param2", "param3");
```
JsDoc @example here as well and look for other example areas.

lnagle replied:
Fixed.

AbsentSemicolon commented on script/test-race-conditions.spec.js:
```diff
@@ -0,0 +1,378 @@
+
+
+/**
+ * Calls a list of methods one after another. This is used for testing that
```
dont need that

lnagle replied:
Fixed.

AbsentSemicolon commented on script/test-race-conditions.spec.js:
```diff
@@ -0,0 +1,378 @@
+ * commands run as expected when they are not called in a reasoned, methodical
+ * order (i.e. one method calls the next in its callback).
+ *
+ * Example:
```
@example

lnagle replied:
Fixed.

AbsentSemicolon commented on script/test-race-conditions.spec.js:
```diff
@@ -0,0 +1,378 @@
+    instance.on("error", (err) => {
+        done(err);
+    });
+    spyOn(instance, "close").andCallThrough();
```
I believe `.andCallThrough()` is the default so you shouldn't need to include it for these.

lnagle replied:
Pretty sure you're wrong. `spyOn` and `createSpy` both make functions that are effectively noops with counters.

AbsentSemicolon replied:
Looks like something that talked about but not implemented in 1.3 at least.

AbsentSemicolon commented on script/test-race-conditions.spec.js:
```diff
@@ -0,0 +1,378 @@
+        ];
+        done = (err) => {
+            expect(err).toBeUndefined();
+            expect(instance.backend.connection).not.toBeUndefined();
```
You could just use `.toBeDefined()`

lnagle replied:
Fixed.

AbsentSemicolon commented on spec/lib/ddq-backend-mysql.spec.js:
```diff
@@ -57,37 +57,44 @@ describe("lib/ddq-backend-mysql", () => {
         });
     });
-    describe(".listen()", () => {
+    describe(".startListeninging()", () => {
```
Think this should be `.startListening()`

lnagle replied:
Fixed.

lnagle committed:
Chris' Code Review Changes

AbsentSemicolon commented:
Looks good, approved.

fidian commented on package.json:
```diff
@@ -27,11 +27,11 @@
     ],
     "license": "MIT",
     "dependencies": {
+        "ddq": "2.0.0",
```
This creates a circular dependency.  You should list this under `peerDependencies`.  https://nodejs.org/en/blog/npm/peer-dependencies/

quantumew replied:
Fixed that. The build will fail though until DDQ is published. Tests are passing locally. My commits 
46eee4607d09d2f88ca247764e170ef043da943b
fb9ac39046289a652a13be072a4d0939a129ed38

quantumew committed:
Committing fix to package json so DDQ can install it.

quantumew committed:
Fix all tests and test missing lines.

fidian commented on spec/lib/ddq-backend-mysql.spec.js:
```diff
@@ -168,33 +168,33 @@ describe("lib/ddq-backend-mysql", () => {
                 callback({
                     code: "ER_DUP_ENTRY"
-                });
+                }, {});
```
Why does the callback take two arguments?

quantumew replied:
My mistake 0633ddef4d1696db618d318cc32312f7d6db5960

fidian commented on spec/lib/ddq-backend-mysql.spec.js:
```diff
@@ -168,33 +168,33 @@ describe("lib/ddq-backend-mysql", () => {
+                    }, {});
                 });
-                instance.sendMessage("Example Message", () => {});
+                instance.sendMessage("Example Message", null, () => {});
```
This should work with and without the `null`.

    instance.sendMessage("Example Message", null, () => {});
    instance.sendMessage("Example Message", () => {});

quantumew replied:
4fe18a097a56af6629bbcf5033de13677b7aec7f

fidian reviewed (changes requested):
Need one thing changed and there's one really weird thing I don't understand.

quantumew committed:
Moved MySql stuff from DDQ and made topic truly optional.

quantumew committed:
Removing unnecessary parameter.

quantumew committed:
Removing empty object and checking data is not null from query.

quantumew commented:
@fidian reverted the topic stuff and added an extra check before checking data.affectedRows so we can remove the empty object from the tests. I see now that was why that was added. 8d914e232f2a7bd04e0ea9483e34b626a0937c09

fidian commented:
Awesome.  Thanks @lnagle93 and @quantumew for your hard work!

fidian commented:
Can't merge this until tests pass.  :-(

quantumew committed:
Correcting path

Wow fixing again.

Fix call to mysql. I don't know what -e was supposed to be.

Why is there a user! Removed

Read travis docs. Looks like root with no password

Updating testing script calls.

Update dependencies

This functional test setup is very annoying.

It is broken but since you need to publish DDQ for this to get the
changes, and then commit fix on travis to even see if it will work.
Also I didn't write this so I am just trying stuff.

Fixing ddq version now that i had to bump to fix test scripts

Shimming version of backend tests.

Add missing argument

Missing MySql field.

Functional tests exposed a bug in a MySql query

Yet another one to fix.

"topics" not "topic" in the MySql setup script.

Ugh should be topic I see now.

Returning empty object if not record is found rather then blowing up.

quantumew commented on package.json:
```diff
@@ -25,18 +25,17 @@
     },
     "devDependencies": {
         "codecov": "^1.0.1",
+        "ddq-backend-tests": "git://github.com/tests-always-included/ddq-backend-tests.git#initial",
```
This is temporary until this is published to get build to pass.

fidian replied:
We shouldn't publish this until the backend tests are published.

quantumew committed:
Don't emit without data.

quantumew committed:
Squashed commit of the following:

commit 48c42f30cc152f92fce0958c08e762fcf4738b7e
Author: Andy Gertjejansen <AndyGertjejansen@outlook.com>
Date:   Fri Jan 20 13:45:43 2017 -0600

    Shimming packagejson

commit 07a77eb5f446d9427cdf7af74874eeb624a3e0dc
Author: Andy Gertjejansen <AndyGertjejansen@outlook.com>
Date:   Fri Jan 20 13:42:37 2017 -0600

    Cosmetic change.

commit 7b643862714431f4f00002ba7d30eff1e495999c
Author: Andy Gertjejansen <AndyGertjejansen@outlook.com>
Date:   Fri Jan 20 13:29:22 2017 -0600

    Use hash for owner and update query to set initial heartbeat date

commit 2250ee7dacef5e288400fdf95095bf662bbc8973
Author: Andy Gertjejansen <AndyGertjejansen@outlook.com>
Date:   Thu Jan 19 12:39:55 2017 -0600

    Remove non-existant script and fix bugs.

    Not much of this plugin was working at all. Owner was incorrect on one
    query and therefore it never worked. The heartbeat cleanup never cleaned
    up because of flawed query. Fix that stuff.

commit ba4961f6a6731b14b6cd4fd5e8a438398361edf1
Author: Andy Gertjejansen <AndyGertjejansen@outlook.com>
Date:   Thu Jan 19 10:23:01 2017 -0600

    Update functional tests and update record before sending.

    When we poll we now update the record to isProcessing = true
    and the owner to config.owner. This appears to have been a pretty major
    oversight. I will be adding functional tests to prevent this.

quantumew commented:
@fidian Here are the plugin bug fixes 0d0ae557415c6f304879562767060f3be0af6dda

Note to reviewers, the package.json is shimmed to point to a branch on ddq-backend-tests to get the tests passing until they are published.

kyle-long commented:
I think we should be careful about the phrase "cleaned up". We use it multiple contexts and sometimes in makes little sense to me. For example...

```
        /**
         * Queries for jobs that have stopped, and primes them to be cleaned up.
         */
        function heartbeatCleanup() {
```

Is it priming them to be "cleaned up" or is it making them consumable again? The record still exist. Cleanup to me implies that we are throwing away that data since we no longer need it.

kyle-long committed:
Adding comment

kyle-long commented on lib/ddq-backend-mysql.js:
```diff
@@ -365,17 +405,17 @@ module.exports = (configValidation, crypto, EventEmitter, mysql, timers) => {
+        sendMessage(message, topic, callback) {
             var hash, params;
 
             cycleCounter = 0;
```
FUNCTIONAL?

sending messages could be done asynchronously. The counter really should be on a per message basis instead of on a module level. For that matter, you could have multiple listeners potentially listening to different topics.

However, how much do we care? The worse that could happen is that it gets stuck in an infinite loop trying to insert the messages if they come in PERFECTLY. The mostly likely scenario nothing happens at all.

quantumew replied:
Fixed

kyle-long commented:
Can't comment on the line because it is not in the diff.

If "remove" is call but the "requeued" column is set to true, it will not call "requeue" in order to set "requeued" to false. Instead nothing will happen.

This is based off of the assumption that a DELETE will not error if it finds no records to delete.

kyle-long commented:
I think the last thing I noticed was if "requeue" was called after the heartbeat expired for that message and a new listener had picked it up it would remove the ability for the new listener to remove that message and instead it would get processed an additional time.

I am not sure how much we care about this one.

kyle-long committed:
fixing up comments

Because I was confused and some things were not properly explained.

quantumew committed:
Make cycleCounter message level so it works in async context.

Also, check the owner in the WHERE clause of the requeue query. You must
own the message to be able to requeue it.

Adding milliseconds to heartbeatDate and remove isProcessing index

With isProcessing indexed we were getting deadlocks. It appears the
index would be scanned and while that happened another query would grab
a lock on a record before the one scanning could. I am still a bit
confused about it tbh.

https://www.xaprb.com/blog/2006/08/03/a-little-known-way-to-cause-a-database-deadlock/

Dealing with deadlocks on restore and polling.

With our current setup deadlocks are always  possible so we have to deal
with them. I added logic to continue on when deadlocks occur during
polling and restoring. I also added a deadlockCountLimit which if
consecutive deadlocks occur it will emit an error once it reaches that
count. Also, update the tests.

quantumew commented:
@kyle-long here are the code review fixes and such 5f28dc90b23f091418f270eeaf79cb0df5af4b43

kyle-long commented on lib/ddq-backend-mysql.js:
```diff
@@ -3,20 +3,23 @@
         ddqBackendInstance.connection.query(`
             UPDATE ??
-            SET heartbeatDate = NOW()
+            SET heartbeatDate = NOW(3)
```
I see, so this makes it more precise for the deadlocking. Could you add a comment about this? After a quick google I couldn't find it here http://www.w3schools.com/sqL/func_now.asp and didn't figure it out until I tested it.

Makes sense though

quantumew replied:
Oh yea good call. Note added. 9e135cb4ce4919b550d88120b2ac35376409cd80

kyle-long commented on script/mysql-commands.sql:
```diff
@@ -0,0 +1,15 @@
+    heartbeatDate DATETIME,
+    owner VARCHAR(256),
+    isProcessing BOOLEAN DEFAULT FALSE,
+    index isProcessing(isProcessing),
```
Why lowercase? I like uppercase and we do it for the rest of the stuff.

fidian replied:
Should be uppercase. https://tests-always-included.github.io/style-guide/lang/sql/
Looks like the SQL guide isn't linked up.  I'll go fix that.

quantumew replied:
Yea, sorry that was not intentional. Fixed 9e135cb4ce4919b550d88120b2ac35376409cd80

quantumew committed:
Add notes about timestamp with milliseconds and fix SQL capitalization.

kyle-long commented:
In those comments I wanted a bit about why we want the time to be more accurate right? Thats because of the deadlocking issue right?

quantumew committed:
More comments about timestamp
q

quantumew committed:
Requeue message if it cannot be removed.

quantumew committed:
Adding debug lines to help with debugging.

quantumew committed:
Remove transaction and select for update.

We now check affectedRows and if there are 0 we do not emit.

quantumew committed:
Tests were not being run by travis and tests were behind.

Also pointing to published tests.

quantumew committed:
Merge branch 'master' into release2.0
